compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -726,10 +726,9 @@ class TypeComparer(initctx: Context) extends DotClass with ConstraintHandling {
 
     tycon2 match {
       case param2: TypeParamRef =>
-        isMatchingApply(tp1) || {
-          if (canConstrain(param2)) canInstantiate(param2)
-          else compareLower(bounds(param2), tyconIsTypeRef = false)
-        }
+        isMatchingApply(tp1) ||
+        canConstrain(param2) && canInstantiate(param2) ||
```

Comment:
I have an alternative fix for #2201  at https://github.com/lampepfl/dotty/pull/2204. Your fix looks like it might help in some other situations, but we would need a testcase for that.